src/__tests__/main.test.ts

```diff
@@ -339,7 +342,121 @@ test("upsert where clause", async (t) => {
+        upsertRole(onConflict: {doUpdate: {name: ignore, updated: current_timestamp}}, where: {
+          projectName: "sales",
+          title: "director"
+        },
+        input: {
+          role: {
+            projectName: "${projectName}",
+            title: "${title}",
+            name: "${name}",
+            rank: ${rank}
+          }
+        }) {
+          clientMutationId
+        }
+      }
+    `);
+    return execGqlOp(t, query);
+  };
+  {
+    // add director
+    await upsertDirector({ name: "jerry" });
+    const res = await fetchAllRoles(t);
+    t.is(res.data.allRoles.edges.length, 1);
+    t.is(res.data.allRoles.edges[0].node.projectName, "sales");
+    t.is(res.data.allRoles.edges[0].node.title, "director");
+    t.is(res.data.allRoles.edges[0].node.name, "jerry");
+    t.is(res.data.allRoles.edges[0].node.updated, null);
+  }
+
+  {
+    // update director
+    await upsertDirector({ name: "frank", rank: 2 });
+    const res = await fetchAllRoles(t);
+    t.is(res.data.allRoles.edges[0].node.projectName, "sales");
+    t.is(res.data.allRoles.edges[0].node.title, "director");
+    t.is(res.data.allRoles.edges[0].node.name, "jerry");
+    t.is(res.data.allRoles.edges[0].node.rank, 1); // rank is unchanged because it is @omit updateOnConflict
+    t.not(res.data.allRoles.edges[0].node.updated, null);
+
+    // assert only one record
+    t.is(res.data.allRoles.edges.length, 1);
+  }
+});
+
+test("upsert where clause on conflict do nothing", async (t) => {
+  const upsertDirector = async ({
+    projectName = "sales",
+    title = "director",
+    name = "jerry",
+    rank = 1,
+  }: {
+    projectName?: string;
+    title?: string;
+    name?: string;
+    rank?: number;
+  }) => {
+    const query = nanographql(`
+      mutation {
+        upsertRole(onConflict: {doNothing: true}, where: {
+          projectName: "sales",
+          title: "director"
+        },
+        input: {
+          role: {
+            projectName: "${projectName}",
+            title: "${title}",
+            name: "${name}",
+            rank: ${rank}
+          }
+        }) {
+          clientMutationId
+        }
+      }
+    `);
+    return execGqlOp(t, query);
+  };
+  {
+    // add director
+    await upsertDirector({ name: "jerry" });
+    const res = await fetchAllRoles(t);
+    t.is(res.data.allRoles.edges.length, 1);
+    t.is(res.data.allRoles.edges[0].node.projectName, "sales");
+    t.is(res.data.allRoles.edges[0].node.title, "director");
+    t.is(res.data.allRoles.edges[0].node.name, "jerry");
+  }
+
+  {
+    // update director
+    await upsertDirector({ name: "frank", rank: 2 });
+    const res = await fetchAllRoles(t);
+    t.is(res.data.allRoles.edges[0].node.projectName, "sales");
+    t.is(res.data.allRoles.edges[0].node.title, "director");
+    t.is(res.data.allRoles.edges[0].node.name, "jerry");
+    t.is(res.data.allRoles.edges[0].node.rank, 1); // rank is unchanged because it is @omit updateOnConflict
```

Comment:
```suggestion
    t.like(
      await upsertDirector({ name: "jerry" }),
      await upsertDirector({ name: "frank", rank: 2 }),
      "onConflict: {doNothing: true} should yield no upsert effect"
    );
```

Reply:
i'm now wondering if this should be an _input_ to the plugin, not the query. think on this, let me know what you think. graphile is all about efficient and secure API/DB operations over GQL, and this implementation leaks backend concerns to the clients. My gut says this all needs to happen on server-side init, not in queries.